### ISSUE TITLE
Increased latency with macOS client 6.0.0

### ISSUE THREAD

**Issue by kostrykin** [moonlight-stream/moonlight-qt#1324]
At first I thought this was related to #1321, but it was pointed out there, that the renderers for macOS and Windows are independent, so I'm opening up a separate issue.

**Describe the bug**
Before 6.0.0, I usually had an "average rendering time" of 0.1-0.2 ms, maybe going up to 0.5 ms at most. After updating to 6.0.0, the "average rendering time" became 2-3 ms, and I could observe a visible lag when moving the mouse connected directly to the host PC (so it feels like input lag, but it is not, since the mouse is connected directly to the host). That amount of observable latency wasn't there with older versions. After downgrading back to 5.0.1, everything is working fine as always again.

**Affected games**
Just streaming the desktop, but it also happens on Counter Strike 2. Seems to be independent of the game.

**Other Moonlight clients**
This does not happen when using older Moonlight versions (e.g., 5.0.1).

**Moonlight settings (please complete the following information)**
I experimented a lot with the settings to find a low-latency configuration which makes CS2 playable:
- Resolution: 1080p @ 120fps
- Video Bitrate: 80 MBps
- Display: Borderless window
- VSync off
- Video decoder: Automatic (recommended)
- Video codec: HEVC (H.265)

**Client PC details (please complete the following information)**
 - OS: macOS 14.5
 - Moonlight Version: v6.0.0
 - Mac Model: iMac (24", M1, 2021)

**Server PC details (please complete the following information)**
 - OS: Windows 10 Education 22H2
 - Sunshine version: 0.21.0
 - GPU: GeForce GTX 1050 Ti
 - GPU driver: 31.0.15.5176

**Moonlight Logs (please attach)**
[Moonlight-1718527238.log](https://github.com/user-attachments/files/15857170/Moonlight-1718527238.log)

**Comment by kostrykin** [comment]
Seems to be resolved in 6.0.1. At least, I don't observe the increased latency that was there in 6.0.0.

**Comment by kostrykin** [comment]
After more testing, I have to amend my previous comment: The increased latency is still there in 6.0.1, compared to 5.0.1.

**Comment by nikber84** [comment]
Confirm. On my Mac M1 - wifi connection
5.0.1 - works without latencies.
6.0.x - every 2 seconds I have small rendering delay

**Comment by heathbarpunch** [comment]
@kostrykin with your latest description of the issue I'm wondering if I'm experiencing the same thing, as I suspect I am, but what I'm experiencing doesn't feel anything like @nikber84's description as I'm not feeling any real _change_ it feels consistent and it feels off.

5.0.1 - Feels great.
6.0.x - feels like I'm moving through molasses, but I attributed it to mouse acceleration changes since my cursor both comes up short and overshoots even on the small figure 8. 

6.0 - [Moonlight-1721960645.log](https://github.com/user-attachments/files/16385825/Moonlight-1721960645.log)
5.1 - [Moonlight-1721960494.log](https://github.com/user-attachments/files/16385816/Moonlight-1721960494.log) 

Same experience optimizing for either remote desktop or gaming. Same experience on WiFi or LAN.

OS: macOS 12.7.4
Mac Model: Macbook Pro 14" M1Pro (Anker TB3 dock)

OS: Windows 10 Pro 22H2
Sunshine version: 0.23.1
GPU: AMD Radeon RX 7800 XT
GPU driver: 24.4.1

**Comment by kostrykin** [comment]
> @kostrykin with your latest description of the issue I'm wondering if I'm experiencing the same thing, as I suspect I am, but what I'm experiencing doesn't feel anything like @nikber84's description as I'm not feeling any real _change_ it feels consistent and it feels off.
> 
> 5.0.1 - Feels great. 6.0.x - feels like I'm moving through molasses, but I attributed it to mouse acceleration changes since my cursor both comes up short and overshoots even on the small figure 8.
> 
> 6.0 - [Moonlight-1721960645.log](https://github.com/user-attachments/files/16385825/Moonlight-1721960645.log) 5.1 - [Moonlight-1721960494.log](https://github.com/user-attachments/files/16385816/Moonlight-1721960494.log)
> 
> Same experience optimizing for either remote desktop or gaming. Same experience on WiFi or LAN.
> 
> OS: macOS 12.7.4 Mac Model: Macbook Pro 14" M1Pro (Anker TB3 dock)
> 
> OS: Windows 10 Pro 22H2 Sunshine version: 0.23.1 GPU: AMD Radeon RX 7800 XT GPU driver: 24.4.1

Yes, for me it's also rather consistent and not the way @nikber84 describes it. Your description "moving through molasses" is much more accurate (feels like I'm moving through molasses).

**Comment by nikber84** [comment]
I've solved my lag problem.

I first thought the problem was with the sixth version of Moonlight. But then microfreezes appeared on the fifth version.
I found a solution to the problem:
https://www.reddit.com/r/GeForceNOW/comments/195l8ff/stuttering_issues_with_geforce_now_on_macos_over/

Now I have both version 5 and 6 working stably.
I don't see any differences

**Comment by kostrykin** [comment]
> I've solved my lag problem.
> 
> I first thought the problem was with the sixth version of Moonlight. But then microfreezes appeared on the fifth version. I found a solution to the problem: https://www.reddit.com/r/GeForceNOW/comments/195l8ff/stuttering_issues_with_geforce_now_on_macos_over/
> 
> Now I have both version 5 and 6 working stably. I don't see any differences

This appears to be a different issue, because I am not playing over Wi-Fi.

**Comment by Write** [comment]
I've tried your settings and mine, both on 6.0.1 and 5.0.1 && 5.0.0, and I have around the same decoding time for both (Around 4ms at 2K@120fps and 2ms with 1080p@120fps).

I noticed that 6.0.X removed the macOS native Game Mode, but I don't have any gain with it anyway. 

I have a M1 Max

**Comment by szc602639921** [comment]
When I am using 6.0.1 version, the latency will go higher (host processing time) when you quickly drag a window left and right. The latency will be higher than 50ms. With 5.0.1, it is fine and the latency is always around 3 ms.
Host system: macos 14.5
remote desktop: Windows 10

**Comment by kostrykin** [comment]
> When I am using 6.0.1 version, the latency will go higher (host processing time) when you quickly drag a window left and right. The latency will be higher than 50ms. With 5.0.1, it is fine and the latency is always around 3 ms. Host system: macos 14.5 remote desktop: Windows 10

This is pretty sure a different issue, since you use macOS as the host system (not the client).

> I noticed that 6.0.X removed the macOS native Game Mode, but I don't have any gain with it anyway.

Could that actually be the reason for the increased latency with 6.0.x clients?

**Comment by kostrykin** [comment]
Issue is still there with 6.1.

**Comment by Danzikks** [comment]
> Проблема с рейтингом 6.1 все еще актуальна.

how did you install moonlight 5.0.1 version? i can't do it, my pc is constantly offline on this version. if i put 6+ my pc appears for streaming....
I don't know what to do. 
What version of Sunshine do you have?

**Comment by kostrykin** [comment]
Are you on macOS 15 maybe? Moonlight 5.0.1 doesn't work anymore on macOS 15 unfortunately, so we are forced to upgrade to 6.1. Since I've upgraded to macOS 15 now, I'm unfortunately not able to compare any more.

**Comment by cgutman** [comment]
Please test the performance and latency of this macOS build. It's using the same libplacebo Vulkan renderer that we use on Linux instead of the existing Metal renderer.

https://nightly.link/moonlight-stream/moonlight-qt/workflows/build/libplacebo_mac

Note: Since the renderer implementations are very different, the rendering latency numbers in the overlay are not comparable to other builds.

**Comment by kostrykin** [comment]
I tested the nightly build.

Streaming with the nightly build _feels_ more smooth than with 6.1.0.

Not sure how reliable the 120fps recording mode of an iPhone is, but I used that to record two videos of my mouse movements and the responses on screen (the output files are 28fps still though). The delays are almost the same between the two versions (4–6 frames at 28fps, i.e. 14–20 ms).

But the strongest difference between the nightly build and 6.1.0 is that with the nightly build the screen is notably brighter. Now that I have seen both versions, I tends to say that it was too dark with 6.1.0. I tried to make two photos for comparison with the same exposure settings with my iPhone:

<img width="2240" height="989" alt="Image" src="https://github.com/user-attachments/assets/14e3842b-2cfb-4e62-ae99-433904079649" />

(the left photo turned out a bit blurry but you can still see the brightness differences, especially in the enlarged image sections)

I will stay on the nightly build for now.

**Comment by cgutman** [comment]
Okay, I've updated the test build to hopefully fix the incorrect colors. It's accessible through the same link ~~(the new build name ends with `f2f250`)~~

Edit: I added another fix for a low FPS issue in full-screen on my M4 Mac Mini, so the latest build is now `f3540b`.

**Comment by cgutman** [comment]
I made several fixes to resolve the reported color issue (06bd8a730bfb2df0382a4fa92467c6d86ca1f783) and improve effective video latency (6e231778a2adcacbe2328142e4aa09b682fa45bc) and now switched the default renderer to libplacebo (e223bf9a1beffdda71df31a22321f45d16c9f3fe), so I will consider this tentatively fixed. If the latest [nightly builds](https://nightly.link/moonlight-stream/moonlight-qt/workflows/build/master) still have problems, please let me know.
